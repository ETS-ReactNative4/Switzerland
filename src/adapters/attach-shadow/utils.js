/**
 * @function attachBoundary
 * ---
 * Takes the node element and attaches the shadow boundary to it if it doesn't exist
 * already. Returns the node if a shadow boundary cannot be attached to the element.
 */
export const attachBoundary = (node, options = {}) => {
    if (node.shadowRoot) return node.shadowRoot;

    const defaultOptions = {
        mode: 'open',
        delegatesFocus: false,
        sheets: null,
    };

    try {
        const boundary = node.attachShadow({
            ...defaultOptions,
            ...options,
            mode: defaultOptions.mode,
        });

        // Attached any adopted stylesheets if any have been passed.
        options.sheets && (boundary.adoptedStyleSheets = [].concat(options.sheets));

        return boundary;
    } catch {
        return node;
    }
};
