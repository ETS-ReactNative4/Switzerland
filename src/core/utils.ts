export function parseName(name: string): [string, CustomElementConstructor, string | null] {
    const [tag, prototype] = name.split('/');
    const extend = prototype ?? null;
    return [tag, getPrototype(prototype), extend];
}

export function getPrototype(name: string): CustomElementConstructor {
    const defaultConstructor = window.document.createElement('div').constructor as CustomElementConstructor;
    if (typeof window === 'undefined') return defaultConstructor;
    return name ? (window.document.createElement(name).constructor as CustomElementConstructor) : defaultConstructor;
}

export function getRandomName(): string {
    return `swiss-${Date.now()}-${Math.round(Math.random() * 100)}`;
}
