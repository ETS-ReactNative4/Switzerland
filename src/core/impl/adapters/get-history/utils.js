import { toCamelcase } from '../../../utils.js';

export function createPatch(getF, types, defaults) {
    return (name) => {
        const key = toCamelcase(name).fromSnake();
        const [f] = [].concat(types[key] || ((a) => a));
        return getF(name) ? f(getF(name)) : defaults[key] || null;
    };
}

export function changeState(f) {
    return (...params) => {
        window.history[f](...params);
        const event = new window.PopStateEvent('popstate');
        window.dispatchEvent(event);
    };
}

export function getParams(types, defaults, location = window.location) {
    const params = new URLSearchParams(location.search);
    const get = params.get.bind(params);
    params.get = createPatch(get, types, defaults);
    return params;
}
