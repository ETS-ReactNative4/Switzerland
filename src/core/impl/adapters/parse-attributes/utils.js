import { toCamelcase } from '../../../utils.js';

export const getAttributes = (attrs, types, defaults) =>
    Object.values(attrs).reduce((acc, attr) => {
        const name = toCamelcase(attr.nodeName).fromKebab();
        const [f] = [].concat(types[name] || ((a) => a));

        return {
            ...acc,
            [name]: f(attr.nodeValue),
        };
    }, defaults);

/**
 * @function hasApplicableMutations
 * ---
 * Only cause a re-render if some of the mutated items have actually changed the attribute when compared, and
 * are not included in the `exclude` list specified in the function's parameters.
 */
export const hasApplicableMutations = (node, mutations, exclude = []) =>
    mutations.some((mutation) => {
        const isObserved = !exclude.includes(mutation.attributeName);
        const isModified = mutation.oldValue !== node.getAttribute(mutation.attributeName);
        return isObserved && isModified;
    });
