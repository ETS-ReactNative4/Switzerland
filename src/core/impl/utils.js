import { consoleMessage, getWindow } from '../../utils.js';
import * as adapters from '../../adapters/index.js';
import * as renderer from '../renderer/index.js';

export async function getInitialProps(node, props) {
    const server = props.server ?? false;

    return {
        node,
        server,
        lifecycle: 'update',
        render: (props) => node?.render?.(props),
        dispatch: dispatchEvent(node),
        window: server ? await getWindow() : window,
        ...props,
    };
}

/**
 * @function dispatchEvent
 * ---
 * Dispatches an event, merging in the current package's version for handling legacy events
 * if/when the payloads differ from version-to-version.
 */
export const dispatchEvent = (node) => (name, payload, options = {}) => {
    const model = typeof payload === 'object' ? payload : { value: payload };

    return node.dispatchEvent(
        new window.CustomEvent(name, {
            bubbles: true,
            composed: true,
            ...options,
            detail: { ...model, version: 4 },
        })
    );
};

async function bindAdapters(renderProps) {
    return Object.entries(adapters).reduce(async (accum, [name, runAdapter]) => {
        const adapters = await accum;
        return { ...adapters, [name]: await runAdapter(renderProps) };
    }, {});
}

function makeCyclicProps(props) {
    props.props = props;
    return props;
}

async function renderTree({ node, boundAdapters, renderProps, runController, runView }) {
    const viewProps = await runController(makeCyclicProps({ ...renderProps, adapter: boundAdapters }));
    const tree = await runView(makeCyclicProps({ ...renderProps, ...viewProps }));
    await renderer.renderTree({ ...renderProps, tree });

    {
        const form = renderer.renderForms(node);

        if (Object.keys(form).length > 0) {
            // Re-render the controller and view if a form is detected in the output.
            const viewProps = await runController(makeCyclicProps({ ...renderProps, adapter: boundAdapters, form }));
            const tree = await runView(makeCyclicProps({ ...renderProps, ...viewProps, form, lifecycle: 'update' }));
            await renderer.renderTree({ ...renderProps, tree, lifecycle: 'update' });
        }
    }
}

export async function runComponent(node, props, [runController, runView]) {
    const renderProps = { ...props, ...(await getInitialProps(node, props)), form: renderer.getForms(node) };
    const boundAdapters = await bindAdapters(renderProps);

    try {
        // Run the controller to gather its props for view rendering.
        return renderTree({ node, boundAdapters, renderProps, runController, runView });
    } catch (error) {
        // Invoke the controller and view again but with passing the error that was thrown.
        process?.env?.NODE_ENV !== 'production' && consoleMessage(error);
        await renderTree({ node, boundAdapters, renderProps: { ...renderProps, error }, runController, runView });

        // Re-throw an error so the caller in `Swiss.render` can clear the queue.
        throw new Error('switzerland/error');
    }
}
