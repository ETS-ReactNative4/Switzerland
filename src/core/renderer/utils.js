import { Swiss } from '../../core/impl/index.js';
import { fromCamelcase } from '../../core/utils.js';
import { getWindow } from '../../utils.js';

const eventListeners = new WeakMap();

export const events = Symbol('switzerland/events');

export function createVNode(name, props = {}, ...children) {
    return {
        name,
        props: typeof name === 'string' ? getProps(props) : props,
        children: children.flat(),
    };
}

export function getProps(props) {
    return Object.entries(props).reduce((props, [key, value]) => {
        if (typeof value === 'object') {
            const mergedProps = Object.entries(value).reduce((props, prop) => {
                const item = getProps({ [`${key}-${prop[0]}`]: prop[1] });
                return { ...props, ...item };
            }, {});

            return { ...props, ...mergedProps };
        }

        return { ...props, [key]: value };
    }, {});
}

export async function getNode(tree, options) {
    const window = await getWindow();

    // Null values should yield to empty strings.
    if (tree == null) return window.document.createTextNode('');

    // Children can be passed through as just string representations.
    if (typeof tree !== 'object') return window.document.createTextNode(String(tree));

    // Delegate to a whole new Swiss custom element.
    if (tree.name instanceof Swiss)
        // Render the nested Swiss component that is effectively a new custom element.
        return tree.name.render(tree.props, options);

    // Otherwise it's a standard element.
    return window.document.createElement(tree.name);
}

export async function getVNodeFragment(...nodes) {
    const window = await getWindow();
    const children = [].concat(nodes.flat());

    // Append all of the nodes to the document fragment.
    const fragment = window.document.createDocumentFragment();
    children.forEach((node) => fragment.appendChild(node));

    return fragment;
}

export function isEvent(key, value) {
    return key.startsWith('on') && typeof value === 'function';
}

export function attachEventListeners(tree) {
    return (node) => {
        // Remove all of the existing event listeners from the node.
        detatchEventListeners(node);

        tree?.props &&
            Object.entries(tree.props).forEach(([key, value]) => {
                if (!isEvent(key, value)) return;

                // Attach all of the required events to the given node.
                const name = fromCamelcase(key.substr(2)).toKebab().substr(1);
                node.addEventListener(name, value);

                // Memorise the events so that can be removed later if necessary.
                !eventListeners.get(node) && eventListeners.set(node, new Set());
                eventListeners.get(node).add({ name, fn: value });
            });
    };
}

export function detatchEventListeners(node) {
    // Remove all of the events currently bound to the node.
    const listeners = eventListeners.get(node) ?? [];

    [...listeners].forEach((listener) => {
        eventListeners.get(node).delete(listener);
        node.removeEventListener(listener.name, listener.fn);
    });
}

export async function getVNodeDOM(tree, options) {
    if (typeof tree.name === 'function') {
        // Gather the sub-tree by invoking the function.
        const subTree = tree.name({ ...tree.props, children: tree.children });

        // If the sub-tree sends us an array then we'll iterate over that to resolve each node.
        if (Array.isArray(subTree))
            return (await Promise.all(subTree.map((tree) => getVNodeDOM(tree, options)))).flat();

        return getVNodeDOM(subTree, options);
    }

    // When passed an array of trees we'll cycle through each one.
    if (Array.isArray(tree)) return (await Promise.all(tree.map((tree) => getVNodeDOM(tree, options)))).flat();

    // Create the node from the given tree.
    const node = await getNode(tree, options);

    // Iterate over each attribute and apply that to the node if it's not a component.
    tree?.props &&
        Object.entries(tree.props).forEach(([key, value]) => {
            if (typeof value === 'function') return;

            if (typeof value === 'boolean') {
                value === true && node.setAttribute(key, '');
                return;
            }

            node.setAttribute(key, value);
        });

    // Iterate over each of the children and yield a node with the HTML content.
    for (const subTree of [].concat(tree?.children ?? [])) {
        const children = await getVNodeDOM(subTree, options);
        [].concat(children).forEach((child) => node.appendChild(child));
    }

    node.attachEventListeners = attachEventListeners(tree);
    node.detatchEventListeners = detatchEventListeners;

    return node;
}
