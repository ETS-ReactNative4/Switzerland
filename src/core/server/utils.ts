import { Child, EventDispatch } from '../../types';
import { getWindow, isNode, isSwiss } from '../../utils';
import { Props, EventName, EventPayload, EventOptions } from '../../types';

export async function getInitialProps({ node, server }: Pick<Props, 'node' | 'server'>): Promise<Props> {
    return {
        node,
        lifecycle: 'mount',
        server,
        render: () => null,
        window: await getWindow(),
        dispatch: dispatch(node),
    };
}

export async function transform(parentNode: HTMLElement, child: Child): Promise<string> {
    const window = await getWindow();

    if (typeof child === 'string') {
        const node = window.document.createTextNode(child);
        parentNode.appendChild(node);
        return parentNode.outerHTML;
    }

    if (isNode(child) || isSwiss(child)) {
        const node = !isSwiss(child.element)
            ? window.document.createElement(child.element)
            : await child.element.render(child.attributes);

        parentNode.appendChild(node);

        if (child.children != null)
            await Promise.all(child.children.map(async (child) => await transform(node, child)));

        return node.outerHTML;
    }

    return '';
}

export function dispatch(node: HTMLElement): EventDispatch {
    return (name: EventName, payload: EventPayload, options: EventOptions = {}) => {
        const model = typeof payload === 'object' ? payload : { value: payload };

        return node.dispatchEvent(
            new window.CustomEvent(name, {
                bubbles: true,
                composed: true,
                ...options,
                detail: { ...model, version: 5 },
            })
        );
    };
}
