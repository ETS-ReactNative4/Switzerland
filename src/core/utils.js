import { handler } from '../middleware/rescue/index.js';
import { styles } from '../middleware/html/utils.js';

/**
 * @constant meta ∷ Symbol
 * ---
 * Used internally by each Switzerland class to store private data, such as the queue and its current state.
 */
export const meta = Symbol('meta');

/**
 * @class Cancel ∷ Error
 * ---
 * Used for the `abort` function which allows the current render of a component to be cancelled, and any
 * queued items to be processed immediately afterwards.
 */
export class Cancel extends Error {}

/**
 * @constant previousProps ∷ HTMLElement e, Props p ⇒ e → p
 * ---
 * Previous props that were used for the last render of each component.
 */
const previousProps = new WeakMap();

/**
 * @constant errorHandlers ∷ HTMLElement e, Props p ⇒ e → p
 * ---
 * An incomplete set of props for a component that are used in the error recovery handler.
 */
const errorHandlers = new WeakMap();

/**
 * @function dispatchEvent ∷ ∀ a b c. HTMLElement e ⇒ e → b → Boolean → Object String c
 * ---
 * Dispatches an event, merging in the current package's version for handling legacy events
 * if/when the payloads differ from version-to-version.
 */
export const dispatchEvent = node => (name, payload, options = {}) => {
    const model = typeof payload === 'object' ? payload : { value: payload };
    return node.dispatchEvent(
        new window.CustomEvent(name, {
            bubbles: true,
            composed: true,
            ...options,
            detail: { ...model, version: 4 }
        })
    );
};

/**
 * @function getRandomId ∷ String
 * ---
 * Uses the browser-native Crypto module for generating a random ID.
 */
export const getRandomId = () => {
    const a = new Uint32Array(1);
    window.crypto.getRandomValues(a);
    return a[0].toString(16);
};

/**
 * @function parseTagName ∷ HTMLElement e ⇒ [String, e]
 * ---
 * Parses the required tag name using the optional forward-slash as a separator for specifying the prototype
 * that the custom element should extend, such as when you're extending a native element. Yields a tuple of
 * a free tag name -- which is determined recursively -- and the prototype that the custom element will extend.
 */
export const parseTagName = name => {
    const parts = name.split('/');
    return [findFreeTagName(parts[0]), determinePrototype(parts[1]), parts[1]];
};

/**
 * @function findFreeTagName ∷ String → String → String
 * ---
 * Resolves the node name by first attempting to use the requested node name, unless it already exists as
 * a custom component. If so, we recursively call the `resolveTagName` function to append a random suffix
 * to the end of the node name until we find a node that isn't registered.
 */
export const findFreeTagName = (name, suffix = null) => {
    const tag = suffix ? `${name}-${suffix}` : name;
    return !window.customElements.get(tag)
        ? tag
        : findFreeTagName(tag, getRandomId());
};

/**
 * @function getEventName ∷ String → String
 * ---
 * Prepends all event names with the '@switzerland' scope.
 */
export const getEventName = label => `@switzerland/${label}`;

/**
 * @function determinePrototype ∷ HTMLElement e ⇒ String → e
 * ---
 * Determines which constructor to extend from for the defining of the custom element. In most cases it
 * will be `HTMLElement` unless the user is extending an existing element.
 */
export const determinePrototype = tag =>
    tag ? document.createElement(tag).constructor : window.HTMLElement;

/**
 * @function consoleMessage ∷ String → String → void
 * ---
 * Takes a message and an optional console type for output. During minification this function will be removed
 * from the generated output if 'NODE_ENV' is defined as 'production', as it will be unused due to 'process.env'
 * checks later on in the code.
 */
export const consoleMessage = (text, type = 'error') =>
    console[type](`\uD83C\uDDE8\uD83C\uDDED Switzerland: ${text}.`);

/**
 * @function initialProps ∷ HTMLElement e, Props p ⇒ e → p → Promise (void) → p
 * ---
 * A utility function for setting all of the initial props that are used for each rendering of a component.
 * Takes the `mergeProps` which a developer can pass to the `render` method.
 */
export const initialProps = (node, mergeProps, scheduledTask) => {
    const utils = {
        dispatch: dispatchEvent(node),
        form: {},
        abort: () => {
            throw new Cancel();
        },
        getLatestProps: () => previousProps.get(node) || null,
        isIdle: () => node[meta].queue.size() <= 1,
        isResolved: async () => {
            const resolution = await Promise.race([
                scheduledTask,
                Promise.resolve(false)
            ]);
            return resolution !== false;
        }
    };

    // Signals are used for temporary data for a single render-pass.
    const prevProps = { ...(previousProps.get(node) || {}) };
    delete prevProps.signal;

    return {
        ...prevProps,
        signal: {},
        ...mergeProps,
        node,
        utils,
        render: node.render.bind(node),
        prevProps: previousProps.get(node),
        lifecycle: mergeProps.lifecycle || 'update'
    };
};

/**
 * @function cycleMiddleware ∷ HTMLElement e, Props p ⇒ e → p → [(p → Promise p|p)] → p
 * ---
 * Iterates over the defined middleware for a component, detecting if any error handlers have been
 * defined, and if so storing the current set of props up to that point. Yields the props that were
 * returned by cycling through each of the middleware functions.
 */
export const cycleMiddleware = async (node, initialProps, middleware) => {
    const props = await middleware.reduce(async (accumP, middlewareP) => {
        const props = { ...(await accumP) };
        props.props = props;
        const middleware = await middlewareP;
        const newProps = middleware(Object.freeze({ ...props }));

        // Determine if there's an error handler in the current set of props. If there is then
        // set the handler function as the default to be used if an error is subsequently thrown.
        handler in newProps && errorHandlers.set(node, newProps);
        return newProps;
    }, initialProps);

    previousProps.set(node, props);
    return props;
};

/**
 * @function handleException ∷ ∀ a. HTMLElement e ⇒ e → Error a → void
 * ---
 * Determines whether an error handler had been set for the component before it errored inside one of
 * the middleware functions. If there's a error handler then it's used -- hopefully to render something to
 * the screen, but it's not guaranteed as it merely works like a typical try-catch. If there is no error
 * handler then `console.error` is used as a last resort.
 */
export const handleException = (node, error) => {
    // Attempt to find an error handler for the current node which can handle the error gracefully.
    // Otherwise a simple yet abrasive `console.error` will be used with no recovery possible.
    const props = errorHandlers.get(node);

    if (!props) {
        consoleMessage(error);
        return;
    }

    previousProps.set(node, { ...props, error });

    return void props[handler]({
        ...props,
        error,
        render: mergeProps => {
            node[meta].state.setNormal();
            return node.render(mergeProps);
        }
    });
};

/**
 * @function cssImports ∷ HTMLElement e ⇒ e → Promise void
 * ---
 * Finds all of the component's `HTMLStyleElement` nodes and extracts the `CSSImportRule` rules, awaiting
 * the resolution of each one before resolving the yielded promise.
 */
export const cssImports = node =>
    styles.has(node) ? Promise.all([...styles.get(node)]) : Promise.resolve();

/**
 * @function getDefaults ∷ ∀ a. Object String (String → a)
 */
export const getDefaults = types =>
    Object.entries(types).reduce(
        (accum, [key, value]) =>
            Array.isArray(value) && typeof value[1] !== 'undefined'
                ? { ...accum, [key]: value[1] }
                : accum,
        {}
    );

/**
 * @function toCamelcase ∷ String → String
 */
export const toCamelcase = value => {
    const f = separator => () => {
        const r = new RegExp(`(${separator}\\w)`, 'g');
        return value.replace(r, match => match[1].toUpperCase());
    };
    return {
        fromKebab: f('-'),
        fromSnake: f('_')
    };
};

/**
 * @function findBoundary ∷ { HTMLElement, ShadowRoot|void } → ShadowRoot|HTMLElement
 */
export const findBoundary = ({ node, boundary }) =>
    boundary || (node[meta] && node[meta].boundary) || node;
