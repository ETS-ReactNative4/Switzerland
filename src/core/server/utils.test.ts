import { getInitialProps } from './utils';

describe('getInitialProps()', () => {
    it('should be able to resolve the initial props', async () => {
        const node = document.createElement('x-imogen');
        const initialProps = await getInitialProps({ node, server: true });

        expect(initialProps.node).toEqual(node);
        expect(initialProps.lifecycle).toEqual('mount');
        expect(initialProps.server).toBeTruthy();
        expect(initialProps.render).toEqual(expect.any(Function));
        expect(initialProps.window).toEqual(expect.any(Object));
        expect(initialProps.dispatch).toEqual(expect.any(Function));
    });
});
