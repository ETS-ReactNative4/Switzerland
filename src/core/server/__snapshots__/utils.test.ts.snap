// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform() should be able to transform simple node hierarchies 1`] = `"<div>Hi Imogen</div>"`;
